controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
 }
 
-func (a *AVRInstance) unpauseSubscription(subscription *subv1.Subscription) error {
-	// Subscription might have been changed since we last read it.  Get the latest...
-	subLookupKey := types.NamespacedName{Name: subscription.Name, Namespace: subscription.Namespace}
-	latestSub := &subv1.Subscription{}
+func (a *AVRInstance) createPVManifestWorkForRestore(newPrimary string) error {
+	pvMWName := BuildManifestWorkName(a.instance.Name, a.instance.Namespace, MWTypePV)
 
-	err := a.reconciler.Get(a.ctx, subLookupKey, latestSub)
-	if err != nil {
-		return fmt.Errorf("failed to retrieve subscription %+v", subLookupKey)
+	existAndApplied, err := a.isManifestExistAndApplied(pvMWName, newPrimary)
+	if err != nil && !errors.IsNotFound(err) {
+		a.log.Error(err, "actuall here")
+
+		return err
 	}
 
-	labels := latestSub.GetLabels()
-	if labels == nil {
-		return fmt.Errorf("failed to find labels for subscription %s", latestSub.Name)
+	if existAndApplied {
+		return nil
 	}
 
-	labels[subv1.LabelSubscriptionPause] = "false"
-	latestSub.SetLabels(labels)
+	a.log.Info("Restore PVs", "newPrimary", newPrimary)
 
-	return a.reconciler.Update(a.ctx, latestSub)
+	return a.restore(newPrimary)
 }
 
-func (a *AVRInstance) processUnpausedSubscription(
-	subscription *subv1.Subscription) (rmn.SubscriptionPlacementDecision, error) {
-	a.log.Info("Processing unpaused Subscription", "name", subscription.Name)
+func (a *AVRInstance) isManifestExistAndApplied(mwName, newPrimary string) (bool, error) {
+	// Try to find whether we have already created a ManifestWork for this
+	mw, err := a.findManifestWork(mwName, newPrimary)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			a.log.Error(err, "Failed to find 'PV restore' ManifestWork")
+
+			return false, nil
+		}
+
+		return false, err
+	}
 
-	homeCluster, peerCluster, err := a.selectPlacementDecision(subscription)
+	if mw != nil {
+		a.log.Info(fmt.Sprintf("Found an existing manifest work (%v)", mw))
+		// Check if the MW is in Applied state
+		if IsManifestInAppliedState(mw) {
+			a.log.Info(fmt.Sprintf("ManifestWork %s/%s in Applied state", mw.Namespace, mw.Name))
+
+			return true, nil
+		}
+
+		a.log.Info("MW is not in applied state", "namespace/name", mw.Namespace+"/"+mw.Name)
+	}
+
+	return false, nil
+}
+
+func (a *AVRInstance) restore(newHomeCluster string) error {
+	// Restore from PV backup location
+	err := a.restorePVFromBackup(newHomeCluster)
 	if err != nil {
-		a.log.Info(fmt.Sprintf("Unable to select placement decision (%v)", err))
+		a.log.Error(err, "Failed to restore PVs from backup")
 
-		return rmn.SubscriptionPlacementDecision{}, err
+		return err
 	}
 
-	if err := a.createOrUpdateVRGRolesManifestWork(homeCluster); err != nil {
-		a.log.Error(err, "failed to create or update VolumeReplicationGroup Roles manifest")
+	return nil
+}
 
-		return rmn.SubscriptionPlacementDecision{}, err
+func (a *AVRInstance) cleanup(fromCluster string) (bool, error) {
+	err := a.deletePVManifestWork(fromCluster)
```

Comment:
Delete PV MW last? Wait for VRG to finish its move to secondary and then delete the PVs?

The PVs are held by the PVC, which is held by VRG, so deleting it here may not have any impact, instead deleting it towards the end may be more useful?

Reply:
Well, I debated whether even deleting the MW for PVs useful.  Maybe we shouldn't even delete the MW for PVs.

Reply:
Will we delete MW for PVs when AVR is deleted, i.e as part of AVR cleanup on delete? If so we can maybe postpone the PV MW deletion. (**need to think more**). A final cleanup is required, otherwise MW seems to act on the assumption that it should keep ManagedCluster upto date with what is in the MW.

We also update PV-MW content when failingover/back based on S3 content, so we additions/removals of PV-MW content would be updated accordingly, hence leaving an MW behind does not seem harmful at present.